app/scripts/controllers/transactions/tx-state-manager.test.js

```diff
@@ -614,6 +631,136 @@ describe('TransactionStateManager', function () {
       );
       assert.equal(result[1].id, 2, 'early txs truncated');
     });
+
+    it('cuts off entire groups of transactions by nonce when adding new transaction', function () {
+      const limit = txStateManager.txHistoryLimit;
+      // In this test case the earliest two transactions are a dropped attempted ether send and a
+      // following cancel transaction with the same nonce. these two transactions should be dropped
+      // together as soon as the 11th unique nonce is attempted to be added, mocked here with limit + 1
+      const txs = generateTransactions(limit + 2, {
+        chainId: currentChainId,
+        to: VALID_ADDRESS,
+        from: VALID_ADDRESS_TWO,
+        nonce: (i) => (i === 1 ? `0` : `${i}`),
+        status: (i) =>
+          i === 0
+            ? TRANSACTION_STATUSES.DROPPED
+            : TRANSACTION_STATUSES.CONFIRMED,
+        type: (i) =>
+          i === 1 ? TRANSACTION_TYPES.CANCEL : TRANSACTION_STATUSES.SENT_ETHER,
+      });
+      txs.forEach((tx) => txStateManager.addTransaction(tx));
+      const result = txStateManager.getTransactions();
+      assert.equal(result.length, limit, `limit of ${limit} is enforced`);
+      assert.notEqual(result[0].id, 0, 'first tx should be removed');
+      assert.notEqual(
+        result[0].status,
+        TRANSACTION_STATUSES.DROPPED,
+        'first tx should not be dropped',
+      );
+      assert.notEqual(
+        result[1].type,
+        TRANSACTION_TYPES.CANCEL,
+        'Cancel tx should be dropped',
+      );
+    });
+
+    it('cuts off entire groups of transactions by nonce + network when adding new transaction', function () {
+      const limit = txStateManager.txHistoryLimit;
+      // In this test case the earliest two transactions are a dropped attempted ether send and a
+      // following cancel transaction with the same nonce. Then, a bit later the same scenario on a
+      // different network. The first two transactions should be dropped after adding even another
+      // single transaction but the other shouldn't be dropped until adding the fifth additional
+      // transaction
+      const txs = generateTransactions(limit + 5, {
+        chainId: (i) => ([0, 1, 4, 5].includes(i) ? currentChainId : '0x1'),
+        to: VALID_ADDRESS,
+        from: VALID_ADDRESS_TWO,
+        nonce: (i) => {
+          if (i === 1) return '0';
+          else if (i === 5) return '4';
```

Comment:
In order to test that we're doing the network grouping correctly, wouldn't we want the two groups on different networks to share the same nonce? That way we can ensure we aren't considering them as one group.

As currently implemented, I think this test would pass if we were grouping by nonce alone.

Reply:
yeah that's a better test. so I should flip the logic so that the [0, 1, 4, 5].includes(i) check returns the same nonce, and this if else returns different networks. I'll do that patch soon

Reply:
that actually matches what I wrote in the description I think? 😕

Reply:
Yeah I think it does!

Reply:
done